src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -1577,240 +1037,237 @@ void hashForModule(String name, byte[] hash) {
-
-            private final Set<T> elements;
-            private final int defaultVarIndex;
-            private final IntSupplier nextLocalVar;
-            private int refCount;
-            private int localVarIndex;
-
-            SetBuilder(Set<T> elements,
-                       int defaultVarIndex,
-                       IntSupplier nextLocalVar) {
-                this.elements = elements;
-                this.defaultVarIndex = defaultVarIndex;
-                this.nextLocalVar = nextLocalVar;
-            }
+                SetReference(Set<T> elements) {
+                    this.elements = sorted(elements);
+                }
 
-            /*
-             * Increments the number of references to this particular set.
-             */
-            final void increment() {
-                refCount++;
+                int increment() {
+                    return ++refCount;
+                }
+
+                int refCount() {
+                    return refCount;
+                }
+
+                List<T> elements() {
+                    return elements;
+                }
+
+                @Override
+                public int compareTo(SetReference<T> o) {
+                    if (o == this) {
+                        return 0;
+                    }
+                    if (elements.size() == o.elements.size()) {
+                        var a1 = elements;
+                        var a2 = o.elements;
+                        for (int i = 0; i < elements.size(); i++) {
+                            var r = a1.get(i).compareTo(a2.get(i));
+                            if (r != 0) {
+                                return r;
+                            }
+                        }
+                        return 0;
+                    } else {
+                        return elements.size() - o.elements.size();
+                    }
+                }
             }
 
             /**
-             * Generate the appropriate instructions to load an object reference
-             * to the element onto the stack.
+             * Build an array to host sets referenced more than once so a given set will only be constructed once.
+             * Transform the bytecode for loading the set onto the operand stack as needed.
              */
-            void visitElement(T element, CodeBuilder cob) {
-                cob.loadConstant((ConstantDesc)element);
-            }
+            class CacheBuilder {
+                private static final String VALUES_ARRAY = "dedupSetValues";
+                final ArrayList<Snippet> cachedValues = new ArrayList<>();
+
+                // Load the set to the operand stack.
+                // When referenced more than once, the value is pre-built with static initialzer
+                // and is load from the cache array with
+                //   dedupSetValues[index]
+                // Otherwise, LoadableSet will load the set onto the operand stack.
```

Comment:
This comment needs update.  No more `LoadableSet`

Reply:
RefCountSet sounds like a set, instead of a ref count. The implementation does hold the sorted elements in a list, but it's still not implementing java.util.Set so I am a little resist the call that a set. Would a RefCountProxy works?

Reply:
Maybe simply `RefCounter`. It's okay for the implementation to hold a sorted list.    The `elements` method can be renamed to `sortedList` that returns a List of elements sorted in ascending order.